File: Pyto/Model/Python Bridging/Selectors/BlockBasedSelector.m
//
//  BlockBasedSelector.m
//
//  Created by Charlton Provatas on 11/2/17.
//  Copyright © 2017 CharltonProvatas. All rights reserved.
//

#import "BlockBasedSelector.h"
#import <objc/runtime.h>

@implementation BlockBasedSelector
@end

void class_addMethodWithBlock(Class class, SEL newSelector, OBJCBlock block)
{
    IMP newImplementation = imp_implementationWithBlock(block);
    Method method = class_getInstanceMethod(class, newSelector);
    class_addMethod(class, newSelector, newImplementation,  method_getTypeEncoding(method));
}

void class_addMethodWithBlockAndSender(Class class, SEL newSelector, OBJCBlockWithSender block)
{
    IMP newImplementation = imp_implementationWithBlock(block);
    Method method = class_getInstanceMethod(class, newSelector);
    class_addMethod(class, newSelector, newImplementation,  method_getTypeEncoding(method));
}

File: Pyto/Python Bridging/Selectors/BlockBasedSelector.swift
//
//  BlockBasedSelector.swift
//  Parking
//
//  Created by Charlton Provatas on 11/9/17.
//  Copyright © 2017 Charlton Provatas. All rights reserved.
//

import Foundation
// swiftlint:disable identifier_name
func Selector(_ block: @escaping () -> Void) -> Selector {
    let selector = NSSelectorFromString("\(CFAbsoluteTimeGetCurrent())")
    class_addMethodWithBlock(_Selector.self, selector) { _ in block() }
    return selector
}

/// used w/ callback if you need to get sender argument
func Selector(_ block: @escaping (Any?) -> Void) -> Selector {
    let selector = NSSelectorFromString("\(CFAbsoluteTimeGetCurrent())")
    class_addMethodWithBlockAndSender(_Selector.self, selector) { (_, sender) in block(sender) }
    return selector
}

// swiftlint:disable identifier_name
let Selector = _Selector.shared
@objc class _Selector: NSObject {
    @objc static let shared = _Selector()
}
